File: tsp-python/src/tsp_core/breaking_change.py
from __future__ import annotations
from dataclasses import dataclass
from .models import (
    TypedSchema, TSPField,
    EnumType, ObjectType,
)


@dataclass(eq=True)
class BreakingChange:
    category: str   # "removed_field" | "required_change" | "type_change" | "enum_value_removed"
    field_path: str
    old: str
    new: str


def detect(old_schema: TypedSchema, new_schema: TypedSchema) -> list[BreakingChange]:
    return _detect_fields(old_schema.fields, new_schema.fields, prefix="")


def _detect_fields(
    old_fields: dict[str, TSPField],
    new_fields: dict[str, TSPField],
    prefix: str,
) -> list[BreakingChange]:
    changes: list[BreakingChange] = []

    for name, old_field in old_fields.items():
        path = f"{prefix}.{name}" if prefix else name

        if name not in new_fields:
            changes.append(BreakingChange(
                category="removed_field",
                field_path=path,
                old=old_field.type.__class__.__name__,
                new="<absent>",
            ))
            continue

        new_field = new_fields[name]

        if not old_field.required and new_field.required:
            changes.append(BreakingChange(
                category="required_change",
                field_path=path,
                old="optional",
                new="required",
            ))

        if type(old_field.type) is not type(new_field.type):
            changes.append(BreakingChange(
                category="type_change",
                field_path=path,
                old=old_field.type.__class__.__name__,
                new=new_field.type.__class__.__name__,
            ))
            continue

        if isinstance(old_field.type, EnumType) and isinstance(new_field.type, EnumType):
            new_values = set(new_field.type.values)
            for v in old_field.type.values:
                if v not in new_values:
                    changes.append(BreakingChange(
                        category="enum_value_removed",
                        field_path=path,
                        old=v,
                        new="<removed>",
                    ))

        if isinstance(old_field.type, ObjectType) and isinstance(new_field.type, ObjectType):
            changes.extend(_detect_fields(old_field.type.fields, new_field.type.fields, prefix=path))

    return changes

File: templane-go/core/breaking_change.go
package core

// BreakingChange describes a schema evolution that breaks existing callers.
type BreakingChange struct {
	Category  string // "removed_field" | "required_change" | "type_change" | "enum_value_removed"
	FieldPath string
	Old       string
	New       string
}

// DetectBreakingChanges compares two schemas and returns all breaking changes.
// Non-breaking: adding optional fields, adding enum values, making required optional.
func DetectBreakingChanges(oldSchema, newSchema *TypedSchema) []BreakingChange {
	var out []BreakingChange
	detectFields(oldSchema.Fields, newSchema.Fields, "", &out)
	return out
}

func detectFields(oldFields, newFields map[string]TemplaneField, prefix string, out *[]BreakingChange) {
	for name, oldField := range oldFields {
		path := name
		if prefix != "" {
			path = prefix + "." + name
		}

		newField, present := newFields[name]
		if !present {
			*out = append(*out, BreakingChange{
				Category:  "removed_field",
				FieldPath: path,
				Old:       oldField.Type.Kind,
				New:       "<absent>",
			})
			continue
		}

		if !oldField.Required && newField.Required {
			*out = append(*out, BreakingChange{
				Category:  "required_change",
				FieldPath: path,
				Old:       "optional",
				New:       "required",
			})
		}

		if oldField.Type.Kind != newField.Type.Kind {
			*out = append(*out, BreakingChange{
				Category:  "type_change",
				FieldPath: path,
				Old:       oldField.Type.Kind,
				New:       newField.Type.Kind,
			})
			continue
		}

		if oldField.Type.Kind == "enum" {
			newValues := make(map[string]bool, len(newField.Type.Values))
			for _, v := range newField.Type.Values {
				newValues[v] = true
			}
			for _, v := range oldField.Type.Values {
				if !newValues[v] {
					*out = append(*out, BreakingChange{
						Category:  "enum_value_removed",
						FieldPath: path,
						Old:       v,
						New:       "<removed>",
					})
				}
			}
		}

		if oldField.Type.Kind == "object" {
			detectFields(oldField.Type.Fields, newField.Type.Fields, path, out)
		}
	}
}
